pkg/transport/generic_transport_controller.go

```diff
@@ -0,0 +1,512 @@
+	if err != nil {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd go
+		// into a loop of attempting to process a work item that is invalid.
+		c.workqueue.Forget(obj)
+		return fmt.Errorf("invalid object key '%s' - %w", key, err)
+	}
+
+	// Run the syncHandler, passing it the PlacementDecision object name to be synced.
+	if err := c.syncHandler(ctx, objectName); err != nil {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.workqueue.AddRateLimited(key)
+		return fmt.Errorf("failed to process object with key '%s' - %w", key, err)
+	}
+	// Finally, if no error occurs we Forget this item so it does not
+	// get queued again until another change happens.
+	c.workqueue.Forget(obj)
+
+	return nil
+}
+
+// syncHandler compares the actual state with the desired, and attempts to converge actual state to the desired state.
+// returning an error from this function will result in a requeue of the given object key.
+// therefore, if object shouldn't be requeued, don't return error.
+func (c *genericTransportController) syncHandler(ctx context.Context, objectName string) error {
+	// Get the PlacementDecision object with this name from WDS
+	placementDecision, err := c.placementDecisionLister.Get(objectName)
+	if err != nil {
+		if errors.IsNotFound(err) { // the object was deleted and it had no finalizer on it. this means transport controller
+			// finished cleanup of wrapped objects from mailbox namespaces. no need to do anything in this state.
+			return nil
+		}
+		// in case of a different error, log it and retry.
+		return fmt.Errorf("failed to get PlacementDecision object '%s' - %w", objectName, err)
+	}
+
+	if isObjectBeingDeleted(placementDecision) {
+		return c.deleteWrappedObjectsAndFinalizer(ctx, placementDecision)
+	}
+	// otherwise, object was not deleted and no error occurered while reading the object.
+	return c.updateWrappedObjectsAndFinalizer(ctx, placementDecision)
+}
+
+// isObjectBeingDeleted is a helper function to check if object is being deleted.
+func isObjectBeingDeleted(object metav1.Object) bool {
+	return !object.GetDeletionTimestamp().IsZero()
+}
+
+func (c *genericTransportController) deleteWrappedObjectsAndFinalizer(ctx context.Context, placementDecision *v1alpha1.PlacementDecision) error {
+	for _, destination := range placementDecision.Spec.Destinations { // TODO need to revisit the destination struct and see how to use it properly
+		if err := c.transportClient.Resource(c.wrappedObjectGVR).Namespace(destination.ClusterId).Delete(ctx, fmt.Sprintf("%s-%s", placementDecision.GetName(), c.wdsName),
+			metav1.DeleteOptions{}); err != nil { // wrapped object name is in the format (PlacementDecision.GetName()-WdsName). see updateWrappedObject func for explanation.
+			if !strings.HasSuffix(err.Error(), notFoundErrorSuffix) { // if object is already not there, we do not report an error cause desired state was achieved.
```

Comment:
Wouldn't it be better to do something with data structure rather than string parsing?

Reply:
I prefer keeping it simple. 
I don't feel like adding data structure would be better in any way.